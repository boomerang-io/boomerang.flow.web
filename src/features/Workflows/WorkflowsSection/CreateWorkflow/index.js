import React, { Component } from "react";
import PropTypes from "prop-types";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import { actions as workflowActions } from "State/workflow";
import { actions as workflowRevisionActions } from "State/workflowRevision";
import { actions as importWorkflowActions } from "State/importWorkflow";
import { ModalFlow, notify, ToastNotification } from "@boomerang/carbon-addons-boomerang-react";
import { Add32 } from "@carbon/icons-react";
import CreateWorkflowContainer from "./CreateWorkflowContainer";
import { BASE_SERVICE_URL } from "Config/servicesConfig";
import DiagramApplication, { createWorkflowRevisionBody } from "Utilities/DiagramApplication";
import styles from "./createWorkflow.module.scss";
import queryString from "query-string";

export class CreateWorkflow extends Component {
  static propTypes = {
    fetchTeams: PropTypes.func.isRequired,
    importWorkflowActions: PropTypes.object.isRequired,
    team: PropTypes.object.isRequired,
    workflowActions: PropTypes.object.isRequired,
    workflowRevisionActions: PropTypes.object.isRequired,
    onCloseModal: PropTypes.bool.isRequired
  };

  diagramApp = new DiagramApplication({ dag: null, isLocked: false });

  createWorkflow = workflowData => {
    const { workflowActions, workflowRevisionActions, fetchTeams } = this.props;
    let workflowId;
    return workflowActions
      .create(`${BASE_SERVICE_URL}/workflow`, workflowData)
      .then(res => {
        workflowId = res.data.id;
        const dagProps = createWorkflowRevisionBody(this.diagramApp, "Create workflow");
        const workflowRevision = {
          ...dagProps,
          workflowId
        };
        fetchTeams();
        return workflowRevisionActions.create(`${BASE_SERVICE_URL}/workflow/${workflowId}/revision`, workflowRevision);
      })
      .then(res => {
        notify(
          <ToastNotification
            kind="success"
            title="Create Workflow"
            subtitle="Successfully created workflow and version"
          />
        );
        this.props.history.push(`/editor/${workflowId}/designer`);
      })
      .catch(err => {
        notify(
          <ToastNotification kind="error" title="Something's wrong" subtitle="Failed to create workflow and version" />
        );
        return Promise.reject();
      });
  };

  handleImportWorkflowCreation = (data, closeModal, team) => {
    const query = queryString.stringify({ update: false, flowTeamId: team.id });
    return this.props.importWorkflowActions
      .post(`${BASE_SERVICE_URL}/workflow/import?${query}`, data)
      .then(() => {
        notify(<ToastNotification kind="success" title="Update Workflow" subtitle="Workflow successfullyupdated" />);
        closeModal();
        this.props.fetchTeams();
      })
      .catch(err => {
        notify(
          <ToastNotification kind="error" title="Something's wrong" subtitle="Failed to create workflow and version" />
        );
      });
  };

  // handleImportWorkflowCreation = (data, closeModal, team) => {
  //   const { workflowActions, workflowRevisionActions, fetchTeams } = this.props;
  //   let workflowId;
  //   const query = queryString.stringify({ update: false, flowTeamId: team.id });
  //   return workflowActions
  //     // .create(`${BASE_SERVICE_URL}/workflow`, data)
  //     .post(`${BASE_SERVICE_URL}/workflow/import?${query}`, data)
  //     .then(res => {
  //       workflowId = res.data.id;
  //       const dagProps = createWorkflowRevisionBody(this.diagramApp, "Create workflow");
  //       const workflowRevision = {
  //         ...dagProps,
  //         workflowId
  //       };
  //       fetchTeams();
  //       return workflowRevisionActions.create(`${BASE_SERVICE_URL}/workflow/${workflowId}/revision`, workflowRevision);
  //     })
  //     .then(res => {
  //       notify(
  //         <ToastNotification
  //           kind="success"
  //           title="Create Workflow"
  //           subtitle="Successfully created workflow and version"
  //         />
  //       );
  //       closeModal();
  //       this.props.history.push(`/editor/${workflowId}/designer`);
  //     })
  //     .catch(err => {
  //       notify(
  //         <ToastNotification kind="error" title="Something's wrong" subtitle="Failed to create workflow and version" />
  //       );
  //       return Promise.reject();
  //     });
  // };

  render() {
    const { team, teams, isCreating } = this.props;
    return (
      <ModalFlow
        modalTrigger={({ openModal }) => (
          <button className={styles.container} onClick={openModal}>
            <Add32 className={styles.addIcon} />
            <p className={styles.text}>Create a new workflow</p>
          </button>
        )}
        confirmModalProps={{
          title: "Close this?",
          children: "Your request will not be saved"
        }}
        modalHeaderProps={{
          title: "Create a new Workflow",
          subtitle: "Get started with these basics, then proceed to designing it out."
        }}
      >
        <CreateWorkflowContainer
          createWorkflow={this.createWorkflow}
          team={team}
          teams={teams}
          isCreating={isCreating}
          handleImportWorkflowCreation={this.handleImportWorkflowCreation}
        />
      </ModalFlow>
    );
  }
}

const mapStateToProps = state => {
  return {
    teams: state.teams.data,
    isCreating: state.workflow.isCreating
  };
};

const mapDispatchToProps = dispatch => {
  return {
    workflowActions: bindActionCreators(workflowActions, dispatch),
    workflowRevisionActions: bindActionCreators(workflowRevisionActions, dispatch),
    importWorkflowActions: bindActionCreators(importWorkflowActions, dispatch)
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CreateWorkflow);
