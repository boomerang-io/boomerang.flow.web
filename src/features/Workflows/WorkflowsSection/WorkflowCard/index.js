import React, { Component } from "react";
import PropTypes from "prop-types";
import axios from "axios";
import { Link, withRouter } from "react-router-dom";
import fileDownload from "js-file-download";
import { Button, OverflowMenu, OverflowMenuItem } from "carbon-components-react";
import {
  notify,
  ToastNotification,
  ModalFlow,
  ConfirmModal,
  TooltipIcon,
} from "@boomerang/carbon-addons-boomerang-react";
import WorkflowWarningButton from "Components/WorkflowWarningButton";
import WorkflowInputModalContent from "./WorkflowInputModalContent";
import WorkflowRunModalContent from "./WorkflowRunModalContent";
import UpdateWorkflow from "./UpdateWorkflow";
import imgs from "Assets/icons";
import { Run20 } from "@carbon/icons-react";
import { BASE_SERVICE_URL } from "Config/servicesConfig";
import styles from "./workflowCard.module.scss";

class WorkflowCard extends Component {
  static propTypes = {
    deleteWorkflow: PropTypes.func.isRequired,
    executeWorkflow: PropTypes.func.isRequired,
    history: PropTypes.object.isRequired,
    setActiveTeam: PropTypes.func.isRequired,
    teamId: PropTypes.string.isRequired,
    workflow: PropTypes.object.isRequired,
  };

  state = {
    isDeleteModalOpen: false,
    isUpdateWorkflowModalOpen: false,
  };

  componentWillUnmount() {
    this.handleOverflowMenuClose();
  }

  executeWorkflow = ({ redirect, properties }) => {
    this.props.executeWorkflow({
      workflowId: this.props.workflow.id,
      redirect,
      properties,
    });
  };

  /**
   * Format properties to be edited in form by Formik. It doesn't work with property notation :(
   * See: https://jaredpalmer.com/formik/docs/guides/arrays#nested-objects
   * This is safe to do because we don't accept "-" characters in property keys
   * @returns {Array}
   */
  formatPropertiesForEdit = () => {
    const { properties = [] } = this.props.workflow;
    return properties.filter((property) => !property.readOnly);
  };

  handleExportWorkflow = (workflow) => {
    notify(<ToastNotification kind="info" title="Export Workflow" subtitle="Your download will start soon." />);
    axios
      .get(`${BASE_SERVICE_URL}/workflow/export/${workflow.id}`)
      .then((res) => {
        const status = res.status.toString();
        if (status.startsWith("4") || status.startsWith("5"))
          notify(<ToastNotification kind="error" title="Export Workflow" subtitle="Something went wrong." />);
        else fileDownload(JSON.stringify(res.data, null, 4), `${workflow.name}.json`);
      })
      .catch((error) => {
        notify(<ToastNotification kind="error" title="Export Workflow" subtitle="Something went wrong." />);
      });
  };

  setActiveTeam = () => {
    const { setActiveTeam, teamId } = this.props;
    setActiveTeam(teamId);
  };

  setActiveTeamAndRedirect = () => {
    const { history, workflow } = this.props;
    this.setActiveTeam();
    history.push(`/editor/${workflow.id}/designer`);
  };

  /* prevent page scroll when up or down arrows are pressed **/
  preventKeyScrolling = (e) => {
    if ([38, 40].indexOf(e.keyCode) > -1) {
      e.preventDefault();
    }
  };

  handleOverflowMenuOpen = () => {
    window.addEventListener("keydown", this.preventKeyScrolling, false);
  };

  handleOverflowMenuClose = () => {
    window.removeEventListener("keydown", this.preventKeyScrolling, false);
  };

  render() {
    const { deleteWorkflow, fetchTeams, history, teamId, workflow } = this.props;
    const menuOptions = [
      {
        itemText: "Edit Workflow",
        onClick: this.setActiveTeamAndRedirect,
        primaryFocus: true,
      },
      {
        itemText: "View Activity",
        onClick: () => history.push(`/activity?page=0&size=10&workflowIds=${workflow.id}`),
      },

      {
        itemText: "Export .json",
        onClick: () => this.handleExportWorkflow(workflow),
      },
      {
        itemText: "Update .json",
        onClick: () => this.setState({ isUpdateWorkflowModalOpen: true }),
      },
      {
        hasDivider: true,
        itemText: "Delete",
        isDelete: true,
        onClick: () => this.setState({ isDeleteModalOpen: true }),
      },
    ];

    const formattedProperties = this.formatPropertiesForEdit();

    return (
      <div className={styles.container}>
        <Link to={`/editor/${workflow.id}/designer`} onClick={this.setActiveTeam}>
          <section className={styles.cardInfo}>
            <div className={styles.cardIconContainer}>
              <img className={styles.cardIcon} src={imgs[workflow.icon ? workflow.icon : "docs"]} alt="icon" />
            </div>
            <div className={styles.cardDescriptionContainer}>
              <h2 className={styles.cardName}>{workflow.name}</h2>
              <p className={styles.cardDescription}>{workflow.shortDescription}</p>
            </div>
          </section>
        </Link>
        <section className={styles.cardLaunch}>
          {Array.isArray(formattedProperties) && formattedProperties.length !== 0 ? (
            <ModalFlow
              modalHeaderProps={{
                title: "Workflow Properties",
                subtitle: "Provide property values for your workflow",
              }}
              modalTrigger={({ openModal }) => (
                <Button iconDescription="Run Workflow" renderIcon={Run20} size="field" onClick={openModal}>
                  Run it
                </Button>
              )}
            >
              <WorkflowInputModalContent executeWorkflow={this.executeWorkflow} inputs={formattedProperties} />
            </ModalFlow>
          ) : (
            <ModalFlow
              composedModalProps={{ containerClassName: `${styles.executeWorkflow}` }}
              modalHeaderProps={{
                title: "Execute Workflow?",
                subtitle: '"Run and View" will navigate you to the workflow exeuction view.',
              }}
              modalTrigger={({ openModal }) => (
                <Button iconDescription="Run Workflow" renderIcon={Run20} size="field" onClick={openModal}>
                  Run it
                </Button>
              )}
            >
              <WorkflowRunModalContent executeWorkflow={this.executeWorkflow} />
            </ModalFlow>
          )}
        </section>
        {workflow.templateUpgradesAvailable && (
          <div className={styles.templatesWarningIcon}>
            <TooltipIcon
              direction="top"
              tooltipText={"New version of a task available! To update, edit your workflow."}
            >
              <WorkflowWarningButton />
            </TooltipIcon>
          </div>
        )}
        <OverflowMenu
          flipped
          ariaLabel="Overflow card menu"
          iconDescription="Overflow menu icon"
          onOpen={this.handleOverflowMenuOpen}
          onClose={this.handleOverflowMenuClose}
          style={{ position: "absolute", right: "0" }}
        >
          {menuOptions.map(({ onClick, itemText, ...rest }, index) => (
            <OverflowMenuItem onClick={onClick} itemText={itemText} key={`${itemText}-${index}`} {...rest} />
          ))}
        </OverflowMenu>
        {this.state.isUpdateWorkflowModalOpen && (
          <UpdateWorkflow
            fetchTeams={fetchTeams}
            onCloseModal={() => this.setState({ isUpdateWorkflowModalOpen: false })}
            teamId={teamId}
            workflowId={workflow.id}
          />
        )}
        {this.state.isDeleteModalOpen && (
          <ConfirmModal
            affirmativeAction={() => {
              deleteWorkflow({ teamId, workflowId: workflow.id });
            }}
            affirmativeButtonProps={{ kind: "danger" }}
            affirmativeText="Delete"
            isOpen={this.state.isDeleteModalOpen}
            negativeAction={() => {
              this.setState({ isDeleteModalOpen: false });
            }}
            negativeText="Cancel"
            onCloseModal={() => {
              this.setState({ isDeleteModalOpen: false });
            }}
            title="Delete Workflow"
          >
            Are you sure you want to delete this workflow? There's no going back from this decision.
          </ConfirmModal>
        )}
      </div>
    );
  }
}

export default withRouter(WorkflowCard);
