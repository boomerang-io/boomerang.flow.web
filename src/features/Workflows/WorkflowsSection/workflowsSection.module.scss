@import "~@boomerang/boomerang-styles/lib/colors";

.container {
  display: flex;
  flex-direction: column;
  width: 100%;
  height: auto;
  margin-top: 1.25rem;
  margin-bottom: 2.75rem;
}

.noWorkflowsMessage {
  margin-top: 1rem;
  color: #697077;
  font-size: 1rem;
  font-weight: 300;
  line-height: 1.25rem;
}

.team {
  margin-left: 1rem;
  font-size: 1.5rem;
  font-weight: 600;
  margin: 0rem;
  letter-spacing: 0.0625rem;
}
.workflows {
  display: flex;
  flex-wrap: wrap;
  margin-left: -0.75rem;
}
