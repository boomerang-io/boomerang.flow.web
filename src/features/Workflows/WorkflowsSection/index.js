import React, { Component } from "react";
import PropTypes from "prop-types";
import { withRouter } from "react-router-dom";
import CreateWorkflow from "./CreateWorkflow";
import WorkflowCard from "./WorkflowCard";
import styles from "./workflowsSection.module.scss";

class WorkflowSection extends Component {
  static propTypes = {
    deleteWorkflow: PropTypes.func.isRequired,
    executeWorkflow: PropTypes.func.isRequired,
    fetchTeams: PropTypes.func.isRequired,
    searchQuery: PropTypes.string.isRequired,
    setActiveTeam: PropTypes.func.isRequired,
    setActiveTeamAndRedirect: PropTypes.func.isRequired,
    team: PropTypes.object.isRequired
  };

  render() {
    const {
      deleteWorkflow,
      executeWorkflow,
      fetchTeams,
      history,
      setActiveTeam,
      setActiveTeamAndRedirect,
      searchQuery,
      team
    } = this.props;

    let workflows = [];
    if (searchQuery) {
      workflows = team.workflows.filter(workflow => workflow.name.toLowerCase().includes(searchQuery.toLowerCase()));
    } else {
      workflows = team.workflows;
    }

    const hasTeamWorkflows = team.workflows.length > 0;

    return (
      <section className={styles.container}>
        <div className={styles.header}>
          <h1 className={styles.team}>{`${team.name} (${workflows.length})`}</h1>
          {!hasTeamWorkflows && (
            <p className={styles.noWorkflowsMessage}>
              This team doesnâ€™t have any Workflows - be the first to take the plunge.
            </p>
          )}
        </div>
        <div className={styles.workflows}>
          {workflows.map(workflow => (
            <WorkflowCard
              deleteWorkflow={deleteWorkflow}
              executeWorkflow={executeWorkflow}
              fetchTeams={fetchTeams}
              key={workflow.id}
              setActiveTeam={setActiveTeam}
              teamId={team.id}
              workflow={workflow}
            />
          ))}
          <CreateWorkflow
            team={team}
            setActiveTeamAndRedirect={setActiveTeamAndRedirect}
            history={history}
            fetchTeams={fetchTeams}
          />
        </div>
      </section>
    );
  }
}

export default withRouter(WorkflowSection);
