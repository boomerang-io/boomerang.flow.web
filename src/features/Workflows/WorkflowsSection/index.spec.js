import React from "react";
import WorkflowsSection from "./index";

const mockfn = jest.fn();

const searchQuery = "test";
const setActiveTeamAndRedirect = mockfn;
const updateWorkflows = mockfn;
const setActiveTeam = mockfn;
const executeWorkflow = mockfn;
const deleteWorkflow = mockfn;
const fetchTeams = mockfn;
const handleImportWorkflow = mockfn;
const closeModal = mockfn;
const team = {
  id: "1234",
  name: "Lucas' team",
  workflows: [
    {
      id: "hQDkX9v",
      name: "lucas-workflow-1",
      description: "blablabla",
      status: "published",
      icon: "utility"
    },
    {
      id: "456",
      name: "lucas-workflow-2",
      description: "blablabla",
      status: "published",
      icon: "utility"
    },
    {
      id: "789",
      name: "lucas-workflow-3",
      description: "blablabla",
      status: "draft",
      icon: "secure"
    }
  ]
};

describe("WorkflowsSection --- Snapshot", () => {
  beforeEach(() => {
    document.body.setAttribute("id", "app");
  });
  it("Capturing Snapshot of WorkflowsSection", () => {
    const { baseElement } = rtlReduxRouterRender(
      <WorkflowsSection
        searchQuery={searchQuery}
        team={team}
        updateWorkflows={updateWorkflows}
        executeWorkflow={executeWorkflow}
        deleteWorkflow={deleteWorkflow}
        setActiveTeam={setActiveTeam}
        setActiveTeamAndRedirect={setActiveTeamAndRedirect}
        importWorkflow={{}}
        fetchTeams={fetchTeams}
        handleImportWorkflow={handleImportWorkflow}
        closeModal={closeModal}
      />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
