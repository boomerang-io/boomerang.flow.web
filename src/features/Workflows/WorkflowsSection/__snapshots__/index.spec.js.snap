// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkflowsSection --- Snapshot Capturing Snapshot of WorkflowsSection 1`] = `
<body
  id="app"
>
  <div>
    <section
      class="container"
    >
      <div
        class="header"
      >
        <h1
          class="team"
        >
          Lucas' team (0)
        </h1>
      </div>
      <div
        class="workflows"
      >
        <button
          class="container"
        >
          <svg
            aria-hidden="true"
            class="addIcon"
            focusable="false"
            height="32"
            preserveAspectRatio="xMidYMid meet"
            style="will-change: transform;"
            viewBox="0 0 32 32"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17 15V7h-2v8H7v2h8v8h2v-8h8v-2h-8z"
            />
          </svg>
          <p
            class="text"
          >
            Create a new workflow
          </p>
        </button>
      </div>
    </section>
  </div>
  <div
    class="ReactModalPortal"
  />
</body>
`;
